### PR TITLE
Async clearing config index

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

For https://github.com/istio/istio/issues/43431

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
```
root@kurator-linux-0001:~/go/src/istio.io/istio/pilot/pkg/xds# benchstat old.txt new.txt 
goos: linux
goarch: amd64
pkg: istio.io/istio/pilot/pkg/xds
cpu: Intel(R) Xeon(R) Gold 6278C CPU @ 2.60GHz
                        │   old.txt    │               new.txt               │
                        │    sec/op    │    sec/op     vs base               │
Cache/key-4                8.737µ ± 1%   8.759µ ±  1%        ~ (p=0.818 n=6)
Cache/insert-4             182.7µ ± 3%   191.6µ ±  9%        ~ (p=0.240 n=6)
Cache/insert_same_key-4   137.91µ ± 5%   97.25µ ± 14%  -29.48% (p=0.002 n=6)
Cache/get-4                8.869µ ± 1%   8.867µ ±  1%        ~ (p=0.818 n=6)
Cache/insert_and_get-4     192.4µ ± 3%   198.0µ ±  4%        ~ (p=0.093 n=6)
geomean                    51.87µ        49.14µ         -5.27%
```

**Comment by howardjohn** [comment]
This feels like a bit of complexity/risk for minimal gain - only 30% in a specific case, but not generalized use cases. I wonder if we should tackle the real root of the issue which I suspect is that we have to use a Lock instead of a rwlock for the LRU. https://github.com/istio/istio/issues/40744 shows not much info - only a mutex profile. We should have them show more info on profiles if we are going to hyper optimize for their case.

Maybe we reconsider use of LRU entirely to avoid locking?

BTW https://dgraph.io/blog/post/caching-in-go/ is fairly interesting. I took a stab at integrating their approach into our cache but it doesn't work to well for the same reason we have issues here with configIndex needing its own lock.

IDK, I feel like we are going down a rabbit hole when its not clear these are actually providing real world benefits or just microbenchmarks. I don't want to block things too much though if we think this will have benefits... Even with this change we still lock just as much as before (in terms of raw number of locks, we spend a bit less time under the lock). Just thinking if we may want to consider bigger changes to really remove lock usage?


wdyt @ramaraochavali @stevenctl ? I am fine with this if others want to proceed

**Comment by hzxuzhonghu** [comment]
I agree the test is far from covering all cases. But I would suggest we look this from a different side, does this patch bring any bad or regression while it brings improvement in explicit case?

**Comment by howardjohn** [comment]
Thinking about it a bit more I am actually less concerned with this than https://github.com/istio/istio/pull/43428. Are they coupled?

That one may be fine too, I am still doing research into the safety part (https://github.com/istio/istio/pull/43428#discussion_r1113320799) .I think you may be right, just want to verify.

> does this patch bring any bad or regression while it brings improvement in explicit case?

Well it is a bit more complex... but not too bad.

**Comment by hzxuzhonghu** [comment]
It has some dependency since the clearing index procedure includes both typeIndex and configIndex. If typeIndex removed, we only need to consider if there is any side-effect of config index

File: pilot/pkg/model/xds_cache.go [hunk 1]
```diff
@@ -164,6 +173,26 @@ func newLru(evictCallback simplelru.EvictCallback[string, cacheValue]) simplelru
...
+	go func() {
+		ticker := time.Tick(5 * time.Second)
+		for {
+			select {
+			case <-ticker:
+				l.mu.Lock()
+				for _, keyConfigs := range l.evictQueue {
+					l.clearIndexes(keyConfigs.key, keyConfigs.dependentConfigs)
+				}
+				l.evictQueue = l.evictQueue[:0] // this is to reuse the underlying memory
```
Comment (howardjohn) [inline pilot/pkg/model/xds_cache.go:186]: minor concern: if we always reuse we may grow this unbounded if we have a big spike? Not sure how much of an issue that is. Is there some way to shrink the cap to some large max (10k or something)?

File: pilot/pkg/model/xds_cache.go [hunk 2]
```diff
@@ -183,25 +212,30 @@ func (l *lruCache) onEvict(k string, v cacheValue) {
...
 func (l *lruCache) updateConfigIndex(k string, dependentConfigs []ConfigHash) {
 	for _, cfg := range dependentConfigs {
 		sets.InsertOrNew(l.configIndex, cfg, k)
 	}
-	l.recordDependentConfigSize()
 }
 
 func (l *lruCache) clearConfigIndex(k string, dependentConfigs []ConfigHash) {
+	c, exists := l.store.Get(k)
+	if exists {
```
Comment (howardjohn) [inline pilot/pkg/model/xds_cache.go:227]: can we add some test of this case? Or do we already have them? 

I am a bit worried due to the fixed 5s timer we may be missing coverage of this in the tests

**Review by howardjohn — COMMENTED** [review]
Overall LGTM, jsut pending the discussion in https://github.com/istio/istio/pull/43428 about typeIndex

**Comment by ramaraochavali** [comment]
@hzxuzhonghu  is it possible to verify this in any large cluster where lot of evictions happen  to see what benefits it has? While this is good it would be good to see the results

**Comment by hzxuzhonghu** [comment]
I donot have such test env, would be glad to help if anyone have a big scale test env

File: pilot/pkg/model/xds_cache.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/xds_cache.go:227]: There is already tests, need to do some update to make It work for async clearing

File: pilot/pkg/model/xds_cache.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/xds_cache.go:186]: There is a three indexed slice which can be use to reslice


https://stackoverflow.com/questions/27938177/golang-slice-slicing-a-slice-with-sliceabc

**Comment by howardjohn** [comment]
> I donot have such test env, would be glad to help if anyone have a big scale test env

fwiw https://github.com/howardjohn/pilot-load/ can do about 10k pods on ~4 cpus. But still does take a bit of setup

**Comment by stevenctl** [comment]
LGTM

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/model/xds_cache.go [hunk 3]
```diff
@@ -146,6 +155,28 @@ func newLru(evictCallback simplelru.EvictCallback[string, cacheValue]) simplelru
...
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				l.mu.Lock()
+				for _, keyConfigs := range l.evictQueue {
+					l.clearConfigIndex(keyConfigs.key, keyConfigs.dependentConfigs)
+				}
+				l.evictQueue = l.evictQueue[:0:1000]
```
Comment (wulianglongrd) [inline pilot/pkg/model/xds_cache.go:170]: `l.evictQueue[:0]` looks better

Comment (hzxuzhonghu) [reply pilot/pkg/model/xds_cache.go:170]: This is to prevent unbounded capacity

**Pull request merged**